File: dsh-mobile-link/lib/tunnel.js
// 隧道核心（01 §4.2，契约 §3 逐条实现）——本插件工作量主体
// 帧格式：11B 大端 ver|type|flags|stream_id|len；len ≤ 16373；flags 恒 0（接收方忽略非 0）；
// 状态机红线（skill §6 / 契约 §3.3）：
//   FIN 半关闭继续读（HTTP 响应体随后返回，全关会静默截断响应体）；
//   终态/未知流迟到帧静默忽略（不回 RST——RST 走 ctl 无序，会形成 RST 风暴）；
//   收到 RST 一律不回；重复/无效 OPEN 静默忽略并计错；回 RST 白名单仅三类（超限新 OPEN / 帧头非法 / 本地写错误）；
//   WINDOW.credit = 自上次 WINDOW 以来实际写入 socket 的增量（固定值会泄漏信用，约 43 轮死锁）；
//   未发送队列 单流 ≤256KiB 全局 ≤512KiB（保活帧与 mux PONG 的排队延迟约束，契约 §3.4.7）；
//   信用 0 停读 socket；FIN 走 data（顺序语义）；PING/PONG 走 ctl 且 stream_id=0；
//   保活 5s PING / 15s 无帧判死（兜底触发器；主触发是 ICE/connectionState，由 rtc.js 上报）。
import net from 'node:net'
import { randomBytes } from 'node:crypto'

export const FRAME_HEADER_SIZE = 11
export const MAX_PAYLOAD = 16373
export const FRAME_VER = 1
export const TYPE = { OPEN: 0x01, DATA: 0x02, FIN: 0x03, RST: 0x04, WINDOW: 0x05, PING: 0x06, PONG: 0x07 }
export const AUTHORITY = '127.0.0.1:13080' // OPEN 一致性断言（契约 §1）

const CREDIT_INITIAL = 256 * 1024 // 每流每方向（§3.4.5）
const WINDOW_THRESHOLD = 64 * 1024 // 累计 ≥64KiB 必发 WINDOW
const PER_STREAM_QUEUE_MAX = 256 * 1024
const GLOBAL_QUEUE_MAX = 512 * 1024
const MAX_STREAMS = 128 // §3.4.2
const KEEPALIVE_MS = 5000
const DEAD_MS = 15000

export function encodeFrame(type, streamId, payload = Buffer.alloc(0)) {
  if (payload.length > MAX_PAYLOAD) throw new Error(`frame len ${payload.length} > ${MAX_PAYLOAD}`)
  const out = Buffer.allocUnsafe(FRAME_HEADER_SIZE + payload.length)
  out.writeUInt8(FRAME_VER, 0)
  out.writeUInt8(type, 1)
  out.writeUInt8(0, 2) // flags 恒 0
  out.writeUInt32BE(streamId >>> 0, 3)
  out.writeUInt32BE(payload.length, 7)
  payload.copy(out, FRAME_HEADER_SIZE)
  return out
}

/** 解码一帧。错误码：BAD_VER（断开）/ LEN_OVERFLOW（对流 RST）/ INCOMPLETE（缓冲）。 */
export function decodeFrame(buf) {
  if (buf.length < FRAME_HEADER_SIZE) return { error: 'INCOMPLETE', need: FRAME_HEADER_SIZE - buf.length }
  const ver = buf.readUInt8(0)
  if (ver !== FRAME_VER) return { error: 'BAD_VER', ver }
  const type = buf.readUInt8(1)
  const flags = buf.readUInt8(2) // 读取但不得据此拒帧
  const streamId = buf.readUInt32BE(3)
  const len = buf.readUInt32BE(7)
  if (len > MAX_PAYLOAD) return { error: 'LEN_OVERFLOW', streamId, len }
  if (buf.length - FRAME_HEADER_SIZE < len) return { error: 'INCOMPLETE', need: FRAME_HEADER_SIZE + len - buf.length }
  return { ver, type, flags, streamId, payload: buf.subarray(FRAME_HEADER_SIZE, FRAME_HEADER_SIZE + len) }
}

/**
 * 插件侧隧道（acceptor）：手机发 OPEN → net.connect(转发目标) → 双向泵。
 * 通道由 rtc.js 建立后 attach()；字节透明红线：绝不吞 DSH 的 2s WS Ping、不做任何 UTF-8 转换。
 */
export class Tunnel {
  constructor({ forwardTarget, log = () => {}, onDead, authority = AUTHORITY }) {
    this.forwardTarget = forwardTarget
    this.log = log
    this.onDead = onDead
    this.authority = authority
    this.streams = new Map() // stream_id → session
    this.errors = 0
    this.dataCh = null
    this.ctlCh = null
    this.lastFrameAt = Date.now()
    this.pingTimer = null
    this.deadTimer = null
  }

  /** werift 通道就绪后挂载（label 识别由 rtc.js 保证：'data' 有序 / 'ctl' 无序） */
  attach(dataCh, ctlCh) {
    this.dataCh = dataCh
    this.ctlCh = ctlCh
    // 入站帧接线（werift onmessage 只做解码与入队级处理，绝不阻塞——skill §6）
    dataCh.onmessage = (ev) => this.onFrame(ev?.data ?? ev)
    ctlCh.onmessage = (ev) => this.onFrame(ev?.data ?? ev)
    dataCh.bufferedAmountLowThreshold = 64 * 1024
    if (dataCh.bufferedAmountLow?.subscribe) dataCh.bufferedAmountLow.subscribe(() => this._resumeReads())
    else if ('onbufferedamountlow' in dataCh) dataCh.onbufferedamountlow = () => this._resumeReads()
    this.pingTimer = setInterval(() => {
      try { this.ctlCh?.send(encodeFrame(TYPE.PING, 0, randomBytes(8))) } catch { /* 通道已关 */ }
    }, KEEPALIVE_MS)
    this.pingTimer.unref?.()
    this.deadTimer = setInterval(() => {
      if (Date.now() - this.lastFrameAt > DEAD_MS) {
        this.log('tunnel dead: 15s 无帧（兜底触发器）')
        this.onDead?.()
      }
    }, 1000)
    this.deadTimer.unref?.()
  }

  detach() {
    clearInterval(this.pingTimer)
    clearInterval(this.deadTimer)
    this.pingTimer = this.deadTimer = null
    for (const st of this.streams.values()) { try { st.socket.destroy() } catch {} }
    this.streams.clear()
    this.dataCh = this.ctlCh = null
  }

  _sendData(buf) { try { this.dataCh?.send(buf) } catch {} }
  _sendCtl(buf) { try { this.ctlCh?.send(buf) } catch {} }
  _globalBuffered() {
    let n = 0
    for (const st of this.streams.values()) n += (st.dataCh?.bufferedAmount ?? 0)
    return n
  }

  /** 手机侧 OPEN 到达：校验 authority 一致性 → 连接转发目标 → 建 session */
  _onOpen(frame) {
    const { streamId, payload } = frame
    if (streamId === 0 || this.streams.has(streamId)) { this.errors++; return } // 无效/重复：静默忽略计错（§3.3 v1.0.5），先于白名单①
    let authority = null
    try { authority = JSON.parse(payload.toString('utf8')).authority } catch { /* payload 非 JSON → 断言失败 */ }
    if (authority !== this.authority) { this.errors++; this.log(`OPEN authority mismatch: ${authority}`); return } // 一致性断言（§1）
    if (this.streams.size >= MAX_STREAMS) { // 白名单①：超限新 OPEN → RST
      this._sendCtl(encodeFrame(TYPE.RST, streamId, Buffer.from('stream limit')))
      return
    }
    const [host, port] = this.forwardTarget.split(':')
    const socket = net.connect(Number(port), host)
    const st = {
      socket, credit: CREDIT_INITIAL, writtenSinceWindow: 0,
      finSent: false, finRecv: false, rst: false, readingPaused: false,
      dataCh: this.dataCh,
    }
    this.streams.set(streamId, st)
    socket.on('data', (chunk) => {
      if (st.rst || st.finSent) return
      if (st.credit <= 0) { st.readingPaused = true; socket.pause(); return } // 信用 0 停读（§3.4.5）
      if (this._globalBuffered() >= GLOBAL_QUEUE_MAX || (this.dataCh?.bufferedAmount ?? 0) >= PER_STREAM_QUEUE_MAX) {
        st.readingPaused = true; socket.pause(); return // 未发送队列上限（§3.4.7）
      }
      for (let off = 0; off < chunk.length; off += MAX_PAYLOAD) {
        const slice = chunk.subarray(off, Math.min(off + MAX_PAYLOAD, chunk.length))
        st.credit -= slice.length
        this._sendData(encodeFrame(TYPE.DATA, streamId, slice))
      }
    })
    socket.on('close', () => {
      this._sendWindow(streamId, st) // 关闭前补齐通告（§3.4.5）
      this.streams.delete(streamId)
      if (!st.finSent && !st.rst) { st.finSent = true; this._sendData(encodeFrame(TYPE.FIN, streamId)) } // FIN 走 data
    })
    socket.on('error', (err) => { // 白名单③：本地 socket 写入错误 → RST（走 ctl）
      st.rst = true
      this.streams.delete(streamId)
      this._sendCtl(encodeFrame(TYPE.RST, streamId, Buffer.from(String(err.code || 'ECONNERR'))))
      this.log(`stream ${streamId} socket error: ${err.code || err.message}`)
    })
  }

  _sendWindow(streamId, st) {
    if (st.writtenSinceWindow <= 0) return
    const delta = st.writtenSinceWindow // 增量语义（§3.4.5），不是固定 64KiB
    st.writtenSinceWindow = 0
    const b = Buffer.alloc(4)
    b.writeUInt32BE(Math.min(delta, 0xFFFFFFFF), 0)
    this._sendCtl(encodeFrame(TYPE.WINDOW, streamId, b))
  }

  _resumeReads() {
    for (const st of this.streams.values()) {
      if (st.readingPaused && st.credit > 0 && this._globalBuffered() < GLOBAL_QUEUE_MAX) {
        st.readingPaused = false
        st.socket.resume()
      }
    }
  }

  /** data/ctl 通道来的原始消息（werift onmessage 回调里只做解码与入队级处理，绝不阻塞——skill §6） */
  onFrame(buf) {
    this.lastFrameAt = Date.now()
    const frame = decodeFrame(buf)
    if (frame.error) {
      if (frame.error === 'BAD_VER') { this.log(`FATAL: frame ver=${frame.ver} ≠ 1 → 断开（§3.1）`); this.onDead?.(); return }
      if (frame.error === 'LEN_OVERFLOW') {
        if (frame.streamId !== undefined) this._sendCtl(encodeFrame(TYPE.RST, frame.streamId, Buffer.from('frame too large'))) // 白名单②
        this.errors++
        return
      }
      this.errors++ // INCOMPLETE：werift 按消息分帧，不应出现；丢弃计错
      return
    }
    const { type, streamId } = frame
    switch (type) {
      case TYPE.OPEN: this._onOpen(frame); return
      case TYPE.DATA: {
        const st = this.streams.get(streamId)
        if (!st || st.rst) return // 终态迟到帧：静默忽略，不回 RST（§3.3）
        st.socket.write(frame.payload, () => {
          // 写入完成回调：累计增量，≥64KiB 发 WINDOW（契约 §3.4.5）
          st.writtenSinceWindow += frame.payload.length
          if (st.writtenSinceWindow >= WINDOW_THRESHOLD) this._sendWindow(streamId, st)
        })
        return
      }
      case TYPE.FIN: {
        const st = this.streams.get(streamId)
        if (!st || st.rst) return
        if (st.finRecv) return // 幂等
        st.finRecv = true
        try { st.socket.end() } catch {} // 半关闭：仅本端写方向；继续读继续转发（红线）
        return
      }
      case TYPE.RST: {
        const st = this.streams.get(streamId)
        if (st) { st.rst = true; try { st.socket.destroy() } catch {}; this.streams.delete(streamId) }
        return // 一律不回 RST（防风暴）
      }
      case TYPE.WINDOW: {
        const st = this.streams.get(streamId)
        if (!st || st.rst) return
        if (frame.payload.length >= 4) {
          st.credit += frame.payload.readUInt32BE(0)
          if (st.readingPaused && st.credit > 0) { st.readingPaused = false; st.socket.resume() }
        }
        return
      }
      case TYPE.PING: this._sendCtl(encodeFrame(TYPE.PONG, 0, frame.payload)); return // PONG 回显，stream_id=0
      case TYPE.PONG: return
      default: this.errors++; return // 未知 type：静默忽略计错，不关隧道（§3.3）
    }
  }
}

File: RemoteDSH-flutter/lib/service/tunnel.dart
// 隧道泵（契约 §3.1–§3.5）—— 自 tools/lib/tunnel-pump.mjs 移植；手机侧用 initiator 模式
// 关键红线（skill §6 / 契约 §3.3）：
//   FIN 半关闭（RawSocket.shutdown(send)）继续读；终态迟到帧静默忽略（不回 RST，防风暴）；
//   收到 RST 一律不回；重复/无效 OPEN 静默忽略并计错；WINDOW.credit = 实际写入 socket 的增量；
//   未发送队列 单流 ≤256KiB 全局 ≤512KiB（bufferedAmount）；信用 0 停读 socket；
//   PING/PONG 走 ctl（stream_id=0）；保活 5s / 15s 无帧判死。
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'package:remotedsh_contract/contract/frames.dart';
import 'package:remotedsh_flutter/rtc/rtc_link.dart';

const String authority = '127.0.0.1:13080'; // 契约 §1 固定值
const int creditInitial = 256 * 1024;
const int windowThreshold = 64 * 1024;
const int perStreamQueueMax = 256 * 1024;
const int globalQueueMax = 512 * 1024;
const int maxStreams = 128;
const int keepaliveMs = 5000;
const int deadMs = 15000;

class _Stream {
  final RawSocket socket;
  int credit;
  int writtenSinceWindow = 0;
  bool finSent = false;
  bool finRecv = false;
  bool rst = false;
  bool readingPaused = false;
  _Stream(this.socket) : credit = creditInitial;
}

class TunnelEndpoint {
  final String mode; // 'initiator'（手机）| 'acceptor'（插件；移植保真用）
  final void Function(String) log;
  final Future<RawSocket?> Function()? connectFactory; // acceptor 专用
  final void Function(String reason)? onDead;

  RtcChannel? dataCh;
  RtcChannel? ctlCh;
  final Map<int, _Stream> streams = {};
  int nextStreamId = 1;
  int errors = 0;
  DateTime lastFrameAt = DateTime.now();
  bool _deadFired = false;

  Timer? _pingTimer;
  Timer? _deadTimer;

  TunnelEndpoint({
    required this.mode,
    required this.log,
    this.connectFactory,
    this.onDead,
  }) {
    _pingTimer = Timer.periodic(const Duration(milliseconds: keepaliveMs), (_) {
      try {
        ctlCh?.send(encodeFrame(FrameType.ping, 0, _pingRand()));
        log('KA: PING sent (ctlCh=${ctlCh != null})');
      } catch (e) {
        log('KA: PING send failed: $e');
      }
    });
    _deadTimer = Timer.periodic(const Duration(seconds: 1), (_) {
      if (!_deadFired &&
          DateTime.now().difference(lastFrameAt).inMilliseconds > deadMs) {
        _deadFired = true;
        log('tunnel dead: no frame for 15s (兜底触发器, 契约 §3.5)');
        onDead?.call('no frame 15s');
      }
    });
  }

  /// 通道就绪后回填（RTC 侧 onDataChannel / createDataChannel 完成后调用）
  void bind(RtcChannel data, RtcChannel ctl) {
    dataCh = data;
    ctlCh = ctl;
    data.onMessage = (Uint8List bytes) => _onChannel(bytes, isCtl: false);
    ctl.onMessage = (Uint8List bytes) => _onChannel(bytes, isCtl: true);
  }

  void close() {
    _pingTimer?.cancel();
    _deadTimer?.cancel();
    _teardownAll('endpoint closed');
  }

  void _teardownAll(String reason) {
    for (final st in streams.values) {
      try {
        st.socket.shutdown(SocketDirection.both);
      } catch (_) {}
      try {
        st.socket.close();
      } catch (_) {}
    }
    streams.clear();
    log('tunnel teardown: $reason');
  }

  // ---- 发起方：为一条已 accept 的本地 RawSocket 开流 ----
  int? openStream(RawSocket socket) {
    if (streams.length >= maxStreams) {
      try {
        socket.close();
      } catch (_) {}
      return null; // §3.4 上限（发起方自觉）
    }
    final id = nextStreamId++;
    final st = _Stream(socket);
    streams[id] = st;
    final open = encodeFrame(
        FrameType.open,
        id,
        Uint8List.fromList(
            '{"authority":"$authority"}'.codeUnits)); // §1 authority 一致性断言
    _sendData(open, st);
    _pumpSocketToChannel(id, st);
    return id;
  }

  void _pumpSocketToChannel(int id, _Stream st) {
    RawSocketEvent? handle(RawSocketEvent ev) {
      if (st.rst || st.finSent) return null;
      switch (ev) {
        case RawSocketEvent.read:
          // 信用控制：信用 0 → 停读（§3.4.5 发送方信用为 0 即暂停读）
          if (st.credit <= 0) {
            st.readingPaused = true;
            st.socket.readEventsEnabled = false; // 停读，等 WINDOW 恢复（防 level-triggered 忙转）
            return null;
          }
          while (true) {
            if (st.credit <= 0) {
              st.readingPaused = true;
              st.socket.readEventsEnabled = false;
              return null;
            }
            final want = st.credit < 16373 * 4 ? st.credit : 16373 * 4;
            final chunk = st.socket.read(want);
            if (chunk == null || chunk.isEmpty) break;
            if (!_emitData(id, st, chunk)) return null; // 队列超限 → 暂停
          }
          return null;
        case RawSocketEvent.readClosed:
          // 本地 socket 半关闭（对端关写）→ FIN（走 data，§3.2/§3.4.8）+ 信用补齐
          finalize(id);
          if (!st.finSent && !st.rst) {
            st.finSent = true;
            try {
              _sendData(encodeFrame(FrameType.fin, id), st);
            } catch (_) {}
          }
          return null;
        case RawSocketEvent.closed:
          finalize(id);
          streams.remove(id);
          return null;
        case RawSocketEvent.write:
          _flushPending(id, st);
          return null;
        default:
          return null;
      }
    }

    st.socket.listen(handle);
    // listen 内部消费事件；无需显式 read 循环 —— RawSocket 事件驱动
  }

  /// 返回 false = 队列超限需暂停（调用方须关读事件）
  bool _emitData(int id, _Stream st, Uint8List chunk) {
    // 未发送队列约束（§3.4.7）：单流 256KiB / 全局 512KiB
    final buffered = dataCh?.bufferedAmount ?? 0;
    if (buffered >= perStreamQueueMax || _globalBuffered() >= globalQueueMax) {
      st.readingPaused = true;
      st.socket.readEventsEnabled = false;
      return false;
    }
    st.credit -= chunk.length;
    for (var off = 0; off < chunk.length; off += maxPayload) {
      final end = (off + maxPayload < chunk.length) ? off + maxPayload : chunk.length;
      final slice = Uint8List.sublistView(chunk, off, end);
      final frame = encodeFrame(FrameType.data, id, slice);
      _sendData(frame, st);
    }
    return true;
  }

  void _sendData(Uint8List frame, _Stream st) {
    try {
      dataCh?.send(frame);
    } catch (_) {}
  }

  void _sendCtl(Uint8List frame) {
    try {
      ctlCh?.send(frame);
    } catch (_) {}
  }

  int _globalBuffered() {
    // Dart 侧统一以 data 通道 bufferedAmount 度量（事件驱动更新）
    return dataCh?.bufferedAmount ?? 0;
  }

  /// 通道 → 本地 socket（两个角色同一份状态机，契约 §3.3 逐行实现）
  void _onChannel(Uint8List bytes, {required bool isCtl}) {
    lastFrameAt = DateTime.now();
    Frame frame;
    try {
      frame = decodeFrame(bytes);
    } on FrameError catch (e) {
      if (e.code == 'BAD_VER') {
        log('FATAL: frame ver != 1 → disconnect (§3.1)');
        if (!_deadFired) {
          _deadFired = true;
          onDead?.call('BAD_VER');
        }
        return;
      }
      if (e.code == 'LEN_OVERFLOW') {
        // stream_id 可解析（≥11B 且 ver 合法）→ 对该流 RST；否则丢弃整帧计错
        if (bytes.length >= 11 && bytes[0] == 1) {
          final dv = ByteData.sublistView(bytes);
          final id = dv.getUint32(3, Endian.big);
          _sendCtl(encodeFrame(FrameType.rst, id,
              Uint8List.fromList('frame too large'.codeUnits)));
        }
        errors++;
      } else {
        errors++; // HEADER_INCOMPLETE / truncated：丢弃计错
      }
      return;
    }
    final id = frame.streamId;
    switch (frame.type) {
      case FrameType.open:
        if (mode != 'acceptor') {
          errors++; // 手机不会收到 OPEN（手机是发起方）
          return;
        }
        if (id == 0 || streams.containsKey(id)) {
          errors++; // 无效/重复 OPEN：静默忽略计错，不回 RST（§3.3 v1.0.5）
          return;
        }
        String? a;
        try {
          a = String.fromCharCodes(frame.payload);
          a = a.startsWith('{"authority":"') ? a.substring(14, a.length - 2) : null;
        } catch (_) {}
        if (a != authority) {
          errors++;
          log('OPEN authority mismatch: $a');
          return;
        }
        if (streams.length >= maxStreams) {
          // 白名单①：超限新 OPEN → RST
          _sendCtl(encodeFrame(
              FrameType.rst, id, Uint8List.fromList('stream limit'.codeUnits)));
          return;
        }
        _acceptConnect(id);
        return;
      case FrameType.data:
        final st = streams[id];
        if (st == null || st.rst) return; // 终态/未知流迟到帧：静默忽略（§3.3 v1.0.4）
        _writeToSocket(id, st, frame.payload);
        return;
      case FrameType.fin:
        final st = streams[id];
        if (st == null || st.rst) return;
        if (st.finRecv) return; // 幂等（§3.3）
        st.finRecv = true;
        try {
          st.socket.shutdown(SocketDirection.send); // 半关闭：只关本端写方向，继续读（红线）
        } catch (_) {}
        _maybeClose(id, st);
        return;
      case FrameType.rst:
        final st = streams[id];
        if (st != null) {
          st.rst = true;
          try {
            st.socket.shutdown(SocketDirection.both);
          } catch (_) {}
          try {
            st.socket.close();
          } catch (_) {}
          streams.remove(id);
        }
        return; // 收到 RST 一律不回 RST（§3.3 防风暴）
      case FrameType.window:
        final st = streams[id];
        if (st == null || st.rst) return;
        if (frame.payload.length >= 4) {
          final dv = ByteData.sublistView(frame.payload);
          final delta = dv.getUint32(0, Endian.big);
          st.credit += delta;
          if (st.readingPaused && st.credit > 0) {
            st.readingPaused = false;
            st.socket.readEventsEnabled = true; // 恢复读
          }
        }
        return;
      case FrameType.ping:
        log('KA: PING recv -> PONG');
        _sendCtl(encodeFrame(FrameType.pong, 0, frame.payload)); // 回显随机数，stream_id=0
        return;
      case FrameType.pong:
        log('KA: PONG recv');
        return;
      default:
        errors++; // 未知 type：静默忽略计错，不关隧道（§3.3）
        return;
    }
  }

  Future<void> _acceptConnect(int id) async {
    final raw = await (connectFactory?.call() ?? Future<RawSocket?>.value(null));
    if (raw == null) return;
    final st = _Stream(raw);
    streams[id] = st;
    _pumpSocketToChannel(id, st);
  }

  void _writeToSocket(int id, _Stream st, Uint8List payload) {
    // 已有积压（内核缓冲满）→ 严格按序追加，禁止直写造成流错乱
    final pending = _pendingWrites[id];
    if (pending != null) {
      final merged = Uint8List(pending.length + payload.length);
      merged.setAll(0, pending);
      merged.setAll(pending.length, payload);
      _pendingWrites[id] = merged;
      st.writtenSinceWindow += payload.length;
      if (st.writtenSinceWindow >= windowThreshold) _sendWindow(id, st);
      st.socket.writeEventsEnabled = true;
      return;
    }
    final n = st.socket.write(payload);
    if (n < payload.length) {
      _pendingWrites[id] = Uint8List.sublistView(payload, n);
      st.socket.writeEventsEnabled = true;
    }
    st.writtenSinceWindow += payload.length;
    if (st.writtenSinceWindow >= windowThreshold) _sendWindow(id, st);
  }

  final Map<int, Uint8List> _pendingWrites = {};

  void _flushPending(int id, _Stream st) {
    final pending = _pendingWrites[id];
    if (pending == null) return;
    final n = st.socket.write(pending);
    if (n < pending.length) {
      _pendingWrites[id] = Uint8List.sublistView(pending, n);
      st.socket.writeEventsEnabled = true;
    } else {
      _pendingWrites.remove(id);
      st.socket.writeEventsEnabled = false;
      _maybeClose(id, st);
    }
  }

  void _maybeClose(int id, _Stream st) {
    if (st.finRecv && _pendingWrites[id] == null) {
      // 双向都关了 → 资源清理
      try {
        st.socket.close();
      } catch (_) {}
    }
  }

  void _sendWindow(int id, _Stream st) {
    if (st.writtenSinceWindow <= 0) return;
    final delta = st.writtenSinceWindow;
    st.writtenSinceWindow = 0;
    final b = Uint8List(4);
    ByteData.sublistView(b).setUint32(0, delta > 0xFFFFFFFF ? 0xFFFFFFFF : delta, Endian.big);
    _sendCtl(encodeFrame(FrameType.window, id, b)); // 增量语义（§3.4.5）
  }

  /// 流关闭前的最终补齐通告（§3.4.5：流关闭前 MUST 把已写入字节数补齐通告）
  void finalize(int id) {
    final st = streams[id];
    if (st == null) return;
    _sendWindow(id, st);
  }
}

/// 本地代理（契约 §3.2）：固定 authority，仅 loopback；隧道未就绪立即 close，不排队
class LocalProxy {
  final int port;
  final void Function(String) log;
  final TunnelEndpoint? Function() endpointOf;
  final bool Function() tunnelReady;

  RawServerSocket? _server;

  LocalProxy({
    required this.port,
    required this.log,
    required this.endpointOf,
    required this.tunnelReady,
  });

  Future<void> start() async {
    if (_server != null) return;
    _server = await RawServerSocket.bind(InternetAddress.loopbackIPv4, port);
    log('local proxy on 127.0.0.1:$port (authority 固定，Host=127.0.0.1:13080)');
    _server!.listen((socket) {
      final remote = socket.remoteAddress;
      if (!remote.isLoopback) {
        try {
          socket.shutdown(SocketDirection.both);
        } catch (_) {}
        try {
          socket.close();
        } catch (_) {}
        return;
      }
      final ep = endpointOf();
      if (!tunnelReady() || ep == null) {
        try {
          socket.shutdown(SocketDirection.both);
        } catch (_) {}
        try {
          socket.close();
        } catch (_) {}
        return; // 立即 close，不排队不缓冲
      }
      final id = ep.openStream(socket);
      if (id == null) {
        try {
          socket.shutdown(SocketDirection.both);
        } catch (_) {}
        try {
          socket.close();
        } catch (_) {}
      }
    }, onError: (Object e) {
      log('proxy error: $e');
    });
  }

  Future<void> stop() async {
    try {
      await _server?.close();
    } catch (_) {}
    _server = null;
  }
}

final _pingRand = _SecureRandom8();

class _SecureRandom8 {
  final _r = DateTime.now().microsecondsSinceEpoch;
  int _c = 0;
  Uint8List call() {
    final b = Uint8List(8);
    // PING nonce 仅要求「回显一致性」；避免每帧开 SecureRandom 的开销
    final v = _r + _c++ * 7919;
    for (var i = 0; i < 8; i++) {
      b[i] = (v >> (i * 8)) & 0xff;
    }
    return b;
  }
}
